### PR TITLE
crypto/merkle: optimize merkle tree hashing

### PR DESCRIPTION
**Pull request by marbar3778**
## Description 

Upstream https://github.com/lazyledger/lazyledger-core/pull/351 to optimize merkle tree hashing 

### Benchmarking:

```
benchmark                                 old ns/op     new ns/op     delta
BenchmarkHashAlternatives/recursive-8     22914         21949         -4.21%
BenchmarkHashAlternatives/iterative-8     21634         21939         +1.41%

benchmark                                 old allocs     new allocs     delta
BenchmarkHashAlternatives/recursive-8     398            200            -49.75%
BenchmarkHashAlternatives/iterative-8     399            301            -24.56%

benchmark                                 old bytes     new bytes     delta
BenchmarkHashAlternatives/recursive-8     19088         6496          -65.97%
BenchmarkHashAlternatives/iterative-8     21776         13984         -35.78%
```

cc @odeke-em @cuonglm

### CODE REVIEW

**Review by odeke-em — APPROVED** [review]
LGTM, thank you @marbar3778!

**Review by cuonglm — APPROVED** [review]

**Review by cmwaters — APPROVED** [review]
LGTM

File: CHANGELOG_PENDING.md [hunk 1]
```diff
@@ -99,6 +99,7 @@ Friendly reminder: We have a [bug bounty program](https://hackerone.com/tendermi
 - [statesync] \#6378 Retry requests for snapshots and add a minimum discovery time (5s) for new snapshots.
 - [node/state] \#6370 graceful shutdown in the consensus reactor (@JayT106)
 - [crypto/merkle] \#6443 Improve HashAlternatives performance (@cuonglm)
+- [crypto/merkle] \#6513 Optmize HashAlternatives (@marbar3778)
```
Comment (odeke-em) [inline CHANGELOG_PENDING.md:102]: Optimize

Comment (marbar3778) [reply CHANGELOG_PENDING.md:102]: thanks

**Pull request merged**